upgrade/v110to111/upgrade.go

```diff
@@ -82,3 +148,409 @@ func doLonghornNodeUpgrade(namespace string, lhClient *lhclientset.Clientset) (e
+		return err
+	}
+	return nil
+}
+
+func updateCSIDaemonSetsLastAppliedTolerationsAnnotation(kubeClient *clientset.Clientset, namespace string) (err error) {
+	daemonsetList, err := kubeClient.AppsV1().DaemonSets(namespace).List(metav1.ListOptions{
+		LabelSelector: labels.Set(types.GetBaseLabelsForSystemManagedComponent()).String(),
+	})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list Longhorn daemonsets for toleration annotation update")
+	}
+	for _, ds := range daemonsetList.Items {
+		dsp := &ds
+		needToUpdate, err := needToUpdateTolerationAnnotation(dsp)
+		if err != nil {
+			return err
+		}
+		if !needToUpdate {
+			continue
+		}
+		appliedTolerations := getNonDefaultTolerationList(dsp.Spec.Template.Spec.Tolerations)
+		appliedTolerationsByte, err := json.Marshal(appliedTolerations)
+		if err != nil {
+			return err
+		}
+		if err := util.SetAnnotation(dsp, types.GetLonghornLabelKey(types.LastAppliedTolerationAnnotationKeySuffix), string(appliedTolerationsByte)); err != nil {
+			return err
+		}
+		if _, err := kubeClient.AppsV1().DaemonSets(namespace).Update(dsp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func updateCSIDeploymentsLastAppliedTolerationsAnnotation(kubeClient *clientset.Clientset, namespace string) (err error) {
+	deploymentList, err := kubeClient.AppsV1().Deployments(namespace).List(metav1.ListOptions{
+		LabelSelector: labels.Set(types.GetBaseLabelsForSystemManagedComponent()).String(),
+	})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list Longhorn deployments for toleration annotation update")
+	}
+
+	for _, dp := range deploymentList.Items {
+		dpp := &dp
+		needToUpdate, err := needToUpdateTolerationAnnotation(dpp)
+		if err != nil {
+			return err
+		}
+		if !needToUpdate {
+			continue
+		}
+		appliedTolerations := getNonDefaultTolerationList(dpp.Spec.Template.Spec.Tolerations)
+		appliedTolerationsByte, err := json.Marshal(appliedTolerations)
+		if err != nil {
+			return err
+		}
+		if err := util.SetAnnotation(dpp, types.GetLonghornLabelKey(types.LastAppliedTolerationAnnotationKeySuffix), string(appliedTolerationsByte)); err != nil {
+			return err
+		}
+		if _, err := kubeClient.AppsV1().Deployments(namespace).Update(dpp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getNonDefaultTolerationList(tolerations []v1.Toleration) []v1.Toleration {
+	result := []v1.Toleration{}
+	for _, t := range tolerations {
+		if !util.IsKubernetesDefaultToleration(t) {
+			result = append(result, t)
+		}
+	}
+	return result
+}
+
+func needToUpdateTolerationAnnotation(obj runtime.Object) (bool, error) {
+	objMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+
+	annos := objMeta.GetAnnotations()
+	if annos == nil {
+		return true, nil
+	}
+
+	_, ok := annos[types.GetLonghornLabelKey(types.LastAppliedTolerationAnnotationKeySuffix)]
```

Comment:
What's lastAppliedTolerations used for again?
Also why is it sufficient to check for just the existence of the the lastAppliedTolerations annotation?
i.e why is it not required todo a value comparison?

Reply:
Suppose that users set `k1=v1:NoSchedule` for toleration setting in Longhorn UI. Then user, set `""` for toleration setting in Longhorn UI. We use `lastAppliedTolerations` to correctly remove only the  `k1=v1:NoSchedule` toleration and don't touch other toleration not added by Longhorn (e.g, automatically added by Kubernetes)

If the `lastAppliedTolerations` annotation already exists, we know that the component is already handled by the new logic. So no need to touch it. The old logic doesn't have `lastAppliedTolerations` annotation.